src/function/aggregate/nested/list.cpp

```diff
@@ -5,49 +5,665 @@
+		}
+	}
+
+	if (result.GetType().InternalType() == PhysicalType::VARCHAR) {
+
+		// append all the child chars to one string
+		string str = "";
+		auto linked_child_list = Load<LinkedList>((data_ptr_t)GetListChildData(segment));
+		while (linked_child_list.first_segment) {
+			auto child_segment = linked_child_list.first_segment;
+			auto data = TemplatedGetPrimitiveData<char>(child_segment);
+			str.append(data, child_segment->count);
+			linked_child_list.first_segment = child_segment->next;
+		}
+		linked_child_list.last_segment = nullptr;
+
+		// use length and (reconstructed) offset to get the correct substrings
+		auto aggr_vector_data = FlatVector::GetData(result);
+		auto str_length_data = GetListLengthData(segment);
+
+		// get the substrings and write them to the result vector
+		idx_t offset = 0;
+		for (idx_t i = 0; i < segment->count; i++) {
+			if (!null_mask[i]) {
+				auto str_length = Load<uint64_t>((data_ptr_t)(str_length_data + i));
+				auto substr = str.substr(offset, str_length);
+				auto str_t = StringVector::AddStringOrBlob(result, substr);
+				((string_t *)aggr_vector_data)[total_count + i] = str_t;
+				offset += str_length;
+			}
+		}
+
+	} else if (result.GetType().id() == LogicalTypeId::LIST) {
+
+		auto list_vector_data = FlatVector::GetData<list_entry_t>(result);
+
+		// get the starting offset
+		idx_t offset = 0;
+		if (total_count != 0) {
+			offset = list_vector_data[total_count - 1].offset + list_vector_data[total_count - 1].length;
+		}
+		idx_t starting_offset = offset;
+
+		// set length and offsets
+		auto list_length_data = GetListLengthData(segment);
+		for (idx_t i = 0; i < segment->count; i++) {
+			auto list_length = Load<uint64_t>((data_ptr_t)(list_length_data + i));
+			list_vector_data[total_count + i].length = list_length;
+			list_vector_data[total_count + i].offset = offset;
+			offset += list_length;
+		}
+
+		auto &child_vector = ListVector::GetEntry(result);
+		auto linked_child_list = Load<LinkedList>((data_ptr_t)GetListChildData(segment));
+		ListVector::Reserve(result, offset);
+
+		// recurse into the linked list of child values
+		BuildListVector(&linked_child_list, child_vector, starting_offset);
+
+	} else if (result.GetType().id() == LogicalTypeId::STRUCT) {
+
+		auto &children = StructVector::GetEntries(result);
+
+		// recurse into the child segments of each child of the struct
+		auto struct_children = GetStructData(segment);
+		for (idx_t child_count = 0; child_count < children.size(); child_count++) {
+			auto struct_children_segment = Load<ListSegment *>((data_ptr_t)(struct_children + child_count));
+			GetDataFromSegment(struct_children_segment, *children[child_count], total_count);
+		}
+
+	} else {
+		auto aggr_vector_data = FlatVector::GetData(result);
+
+		// set values
+		for (idx_t i = 0; i < segment->count; i++) {
+			if (aggr_vector_validity.RowIsValid(total_count + i)) {
+				GetPrimitiveDataValue(segment, result.GetType(), aggr_vector_data, i, total_count + i);
```

Comment:
We usually interpret the type outside of the loop. Here, you are looping over the segment, and calling `GetPrimitiveDataValue`, which has a type switch, for each value.

So, effectively we have to go through the type switch once for each value. Even though it's the same type throughout the loop. Could you perhaps create a `TemplatedGetDataFromSegment` function? Then you can interpret the type once in this function (GetDataFromSegment), and then call `TemplatedGetDataFromSegment` with the right template, to get a more efficient loop.

Reply:
To add to that, what would be even nicer if we do not have to do this type switch at all in `GetDataFromSegment`, but instead have various different functions (`GetDataFromListSegment`, `GetDataFromPrimitiveSegment`, ...). Which function to call can then be decided once during bind time, and stored in the `FunctionData`. This can be stored potentially recursively, e.g.:

```cpp
typedef void (*get_data_from_segment_t)(ListSegment *segment, Vector &result, idx_t &total_count);

struct ListBindFunction {
    // raw segment function
    get_data_from_segment_t segment_function;

     // used for structs/lists to store child function pointers 
    vector< ListBindFunction> child_functions;
};

```